### PR TITLE
#1693: Refactor versions listing for packages

### LINKED ISSUE #1693: Refactor versions
**Component**: build 

**Why this is needed**: We need it to have a code more readable and to have a better Debian integration

**What should be done**: refactor version.py in the buildchain

**Implementation proposal** (strongly recommended):

**Test plan**:

### PR DESCRIPTION
**Pull request by jbertran**
**Component**:

**Context**: We're adding a new OS, so we need OS-aware versions definition.

**Summary**:

The goal is to provide an authoritative listing of required packages that it as easy as possible to manage.

```
# Common packages are packages for which we need not care about OS-specific
# divergences.
#
# In this case, either:
#   * the _latest_ version is good enough, and will be the one
#     selected by the package managers (so far: apt and yum).
#   * we have strict version requirements that span OS families, and the
#     versioning schemes _and_ package names do not diverge
#
# Strict version requirements are notably:
#   * kubelet and kubectl which _make_ the K8s version of the cluster
#   * salt-minion which _makes_ the Salt version of the cluster
#
# These common packages may be overridden by OS-specific packages if package
# names or versioning conventions diverge.
#
# Packages that we build ourselves require a version and release as part of
# their build process.
```

**Acceptance criteria**: tests pass, debian files are generated with the correct versions.

---

Closes: #1693

### CODE REVIEW

File: buildchain/buildchain/versions.py [hunk 1]
```diff
@@ -363,6 +363,106 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
 PACKAGES_MAP = {pkg.name: pkg for pkg in RPM_PACKAGES}
 
 
+DEB_PACKAGES = (
```
Comment (slaperche-scality) [inline buildchain/buildchain/versions.py]: This commit should probably be removed (`DEB_PACKAGES` is already defined l.466)

File: buildchain/buildchain/packaging.py [hunk 2]
```diff
@@ -324,18 +324,27 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
...
+    }
+
 # Store these versions in a dict to use with doit.tools.config_changed
-_TO_DOWNLOAD_CONFIG : Dict[str, str] = {
-    pkg.name: "{p.version}-{p.release}".format(p=pkg)
-    for pkg in versions.RPM_PACKAGES
-    if pkg.name not in _RPM_TO_BUILD_PKG_NAMES
-}
+_TO_DOWNLOAD_RPM_CONFIG : Dict[str, str] = _list_packages_to_download(
+    versions.PACKAGES,
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: minor: can be wrapped:

```python
_TO_DOWNLOAD_RPM_CONFIG : Dict[str, str] = _list_packages_to_download(
    versions.PACKAGES, _RPM_TO_BUILD_PKG_NAMES
)
```

ditto for `_TO_DOWNLOAD_DEB_CONFIG`

File: buildchain/buildchain/packaging.py [hunk 3]
```diff
@@ -324,18 +324,27 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
...
+    _RPM_TO_BUILD_PKG_NAMES
+)
 
-_TO_DOWNLOAD_DEB_CONFIG : Dict[str, str] = {
-    pkg.name: "{p.version}-{p.release}".format(p=pkg)
-    for pkg in versions.DEB_PACKAGES
-    if pkg.name not in _DEB_TO_BUILD_PKG_NAMES
-}
+# Store these versions in a dict to use with doit.tools.config_changed
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py:331]: Not sure we want to duplicate this comment

File: buildchain/buildchain/packaging.py [hunk 4]
```diff
@@ -308,12 +308,32 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
...
+            sources=SCALITY_RPM_REPOSITORY.get_rpm_path(CALICO_RPM)
+        ),
+    )
+}
+
+def _list_packages_to_build(
+    pkg_cats: Dict[str, Tuple[targets.Package, ...]]
+) -> List[str]:
+    return [
+        pkg.name for pkg_list in pkg_cats.values()
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: Can be squashed:

```python
    return [
        pkg.name for pkg_list in pkg_cats.values() for pkg in pkg_list
    ]
```

File: buildchain/buildchain/packaging.py [hunk 5]
```diff
@@ -202,6 +227,12 @@ def task__build_rpm_repositories() -> Iterator[types.TaskDict]:
     for repository in RPM_REPOSITORIES:
         yield from repository.execution_plan
 
+
+# }}}
+
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: Remove empty lines between folds

File: buildchain/buildchain/packaging.py [hunk 6]
```diff
@@ -239,6 +270,10 @@ def task__build_rpm_repositories() -> Iterator[types.TaskDict]:
     task_dep=['_build_root'],
 )
 
+# }}}
+
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: ditto

File: buildchain/buildchain/packaging.py [hunk 7]
```diff
@@ -381,6 +349,65 @@ def _list_packages_to_download(
     task_dep=['_package_mkdir_rpm_iso_root'],
 )
 
+# }}}
+
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: ditto

File: buildchain/buildchain/packaging.py [hunk 8]
```diff
@@ -328,8 +327,8 @@ def _rpm_package(name: str, sources: List[Path]) -> targets.RPMPackage:
 # All packages not referenced in `RPM_TO_BUILD` but listed in
 # `versions.RPM_PACKAGES` are supposed to be downloaded.
 RPM_TO_DOWNLOAD : FrozenSet[str] = frozenset(
-    "{p.name}-{p.version}-{p.release}".format(p=package)
-    for package in versions.RPM_PACKAGES
+    package.rpm_full_name
+    for package in versions.PACKAGES
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: Why is `RPM_PACKAGES` renamed `PACKAGES`?

File: buildchain/buildchain/versions.py [hunk 9]
```diff
@@ -364,200 +425,65 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
 
 
 DEB_PACKAGES = (
-    Package(
+    PackageVersion(
```
Comment (slaperche-scality) [inline buildchain/buildchain/versions.py]: Save some screen real estate and squash into a single line: `PackageVersion(name='containerd'),` (ditto for the others)

**Review by slaperche-scality — CHANGES_REQUESTED** [review]

File: buildchain/buildchain/versions.py (hunk 9 above)
Comment (jbertran) [reply buildchain/buildchain/versions.py]: Done in a following commit

File: buildchain/buildchain/versions.py (hunk 1 above)
Comment (jbertran) [reply buildchain/buildchain/versions.py]: Oops, indeed. Rebase gone wrong

File: buildchain/buildchain/packaging.py (hunk 8 above)
Comment (jbertran) [reply buildchain/buildchain/packaging.py]: Gah, stray rebase conflict resolution, thank you

File: buildchain/buildchain/packaging.py [hunk 10]
```diff
@@ -1,6 +1,4 @@
 # coding: utf-8
-
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: Why?

Also, it become inconsistent with the other files.

File: buildchain/buildchain/versions.py [hunk 11]
```diff
@@ -228,132 +228,193 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
...
+        Arguments:
+            name: the name of the package
+            version: the version of the package
+            release: the release of the package
+        """
+        self._name = name
+        self._version = version
+        self._release = release
+
+    name = property(operator.attrgetter('_name'))
```
Comment (slaperche-scality) [inline buildchain/buildchain/versions.py:261]: minor: would align

```
    name    = property(operator.attrgetter('_name'))
    version = property(operator.attrgetter('_version'))
    release = property(operator.attrgetter('_release'))
```

File: buildchain/buildchain/versions.py [hunk 12]
```diff
@@ -228,132 +228,193 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
...
+        if self.version:
+            full_version = self.version
+            if self.release:
+                full_version = '{}-{}'.format(self.version, self.release)
+        return full_version
+
+    @property
+    def rpm_full_name(self) -> str:
+        """The package's full name in RPM conventions."""
+        name_fmt = '{p.name}'
```
Comment (slaperche-scality) [inline buildchain/buildchain/versions.py]: could it be rewritten:

```python
if self.version:
    return '{}-{}'.format(self.name, self.full_version)
return self.name
```
?

File: buildchain/buildchain/versions.py [hunk 13]
```diff
@@ -228,132 +228,193 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
...
+        if self.version:
+            name_fmt += '-{p.version}'
+            if self.release:
+                name_fmt += '-{p.release}'
+        return name_fmt.format(p=self)
+
+    @property
+    def deb_full_name(self) -> str:
```
Comment (slaperche-scality) [inline buildchain/buildchain/versions.py:284]: ditto, could we do something like:

```
if self.version:
    return '{}={}'.format(self.name, self.full_version)
return self.name
```
?

File: buildchain/buildchain/versions.py [hunk 14]
```diff
@@ -291,201 +294,113 @@ def deb_full_name(self) -> str:
...
-    ),
-)
+    Arguments:
+        os_family: OS_family for which to list packages
+    """
+    common_pkgs = PACKAGES['common']
+    os_family_pkgs = PACKAGES.get(os_family)
+
+    if os_family_pkgs is None:
+        raise Exception('Unlisted OS family: {}'.format(os_family))
```
Comment (slaperche-scality) [inline buildchain/buildchain/versions.py]: s/Unlisted/Unsupported/?

File: buildchain/buildchain/packaging.py [hunk 15]
```diff
@@ -52,7 +52,7 @@ def _list_packages_to_build(
 def _list_packages_to_download(
     package_versions: Tuple[versions.PackageVersion, ...],
     packages_to_build: List[str]
-) -> Dict[str,str]:
+) -> Mapping[str, Optional[str]]:
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: Why the change here? (and BTW, why the change in general: a commit message would be welcome).

IIRC, the good practice is to use "precise" type for return type and "abstract" one for input.

See https://docs.python.org/3/library/typing.html#typing.Dict:

> Useful for annotating return types. To annotate arguments it is preferred to use an abstract collection type such as Mapping.

File: buildchain/buildchain/packaging.py [hunk 16]
```diff
@@ -346,6 +339,61 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
     task_dep=['_package_mkdir_rpm_iso_root'],
 )
 
+# }}}
+# Debian packages and repositories {{{
+
+def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
+    try:
+        pkg_info = versions.RPM_PACKAGES_MAP[name]
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py]: `_deb_package` using `RPM_PACKAGES_MAP`?

**Review by slaperche-scality — CHANGES_REQUESTED** [review]

File: buildchain/buildchain/packaging.py (hunk 16 above)
Comment (jbertran) [reply buildchain/buildchain/packaging.py]: Oh boy that's bad.

File: buildchain/buildchain/versions.py (hunk 13 above)
Comment (jbertran) [reply buildchain/buildchain/versions.py:284]: Oh indeed, that's basic DRY. Thanks.

File: buildchain/buildchain/versions.py (hunk 14 above)
Comment (jbertran) [reply buildchain/buildchain/versions.py]: Eeeeh technically we don't support Ubuntu yet but we download its packages. As `versions` is supposed to be absolutely standalone I'd rather observe things that pertain to it and only to it, hence unlisted in the context of the package dict keys.

File: buildchain/buildchain/versions.py (hunk 11 above)
Comment (jbertran) [reply buildchain/buildchain/versions.py:261]: Damn that triggers my linter like crazy. Done.

File: buildchain/buildchain/versions.py (hunk 14 above)
Comment (slaperche-scality) [reply buildchain/buildchain/versions.py]: Make sense, but I think we can have a better word than `unlisted`.
Maybe `no packages for OS family {}`?

Comment (jbertran) [reply buildchain/buildchain/versions.py]: Suits me, thanks

File: buildchain/buildchain/packaging.py [hunk 17]
```diff
@@ -27,7 +27,7 @@
 
 
 from pathlib import Path
-from typing import Dict, FrozenSet, Iterator, List, Tuple
+from typing import Dict, FrozenSet, Iterator, List, Mapping, Optional, Tuple
```
Comment (slaperche-scality) [inline buildchain/buildchain/packaging.py:30]: Still no explanation on why this switch in the commit message

**Review by slaperche-scality — APPROVED** [review]
